akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala

```diff
@@ -2404,6 +2404,43 @@ trait FlowOps[+Out, +Mat] {
    */
   def zip[U](that: Graph[SourceShape[U], _]): Repr[(Out, U)] = via(zipGraph(that))
 
+  /**
+   * Combine the elements of current flow and the given [[Source]] into a stream of tuples.
+   *
+   * '''Emits when''' at first emits when both inputs emit, and then as long as any input emits (coupled to the default value of the completed input).
+   *
+   * '''Backpressures when''' downstream backpressures
+   *
+   * '''Completes when''' all upstream completes
+   *
+   * '''Cancels when''' downstream cancels
+   */
+  def zipAll[U, A >: Out](that: Graph[SourceShape[U], _], thisElem: A, thatElem: U): Repr[(A, U)] = {
+    via(zipAllFlow(that, thisElem, thatElem))
+  }
+
+  protected def zipAllFlow[U, A >: Out, Mat2](
+      that: Graph[SourceShape[U], Mat2],
+      thisElem: A,
+      thatElem: U): Flow[Out @uncheckedVariance, (A, U), Mat2] = {
+    case object passedEnd
+    val passedEndSrc = Source.repeat(passedEnd)
+    val left: Flow[Out, Any, NotUsed] = Flow[A].concat(passedEndSrc)
+    val right: Source[Any, Mat2] = Source.fromGraph(that).concat(passedEndSrc)
```

Comment:
Instead of cycling `passedEnd` element, here we could concat the flows with `Source.repeat(thisElem)` and `Source.repeat(thatElem)` respectively. Then the `map` operator below that converts `passedEnd` to `this/thatElem` would not be needed anymore.

Reply:
In this case we will not be able to distinguish between an actual eof marker or a value from the original stream. It's quite possible that the provided default is also a valid value for stream elements.

Reply:
I suspect the problem with that is when to complete the resulting stream: we might encounter `thisElem, thatElem` before both streams have reached their end.

Reply:
Ah yes. That makes sense!